### ISSUE TITLE
Is there a bug in TEST(multi_layer, gradient_check)

### ISSUE THREAD

**Issue by dianyancao** [nyanp/tiny-cnn#60]
<p>I run the test project,failed to pass  TEST(multi_layer, gradient_check) use the following code.</p>
<p>When the size of the hidden layer of the multi_layer network is small it works,but fails when the size greater than 100.
</p>

<pre><code>
TEST(multi_layer, gradient_check) { // sigmoid - cross-entropy
    typedef cross_entropy loss_func;
    typedef sigmoid activation;
    typedef network<loss_func, gradient_descent_levenberg_marquardt> network;

    network nn;
/*
nn << fully_connected_layer<activation>(3, 100)
<< fully_connected_layer<activation>(100, 2);
	*/
//above can pass the gradient check,but the fllowing can't
	nn << fully_connected_layer<activation>(3, 101)
		<< fully_connected_layer<activation>(101, 2);

    vec_t a(3, 0.0);
    label_t t = 1;

    uniform_rand(a.begin(), a.end(), 0, 1);
    nn.init_weight();
    EXPECT_TRUE(nn.gradient_check(&a, &t, 1, 1e-4, GRAD_CHECK_ALL));
}

</code></pre>

**Comment by dianyancao** [comment]
<p>I have printed out the bprop and numerical gradient by adding the code before abs function </p>
<p>in float_t calc_delta(const vec_t* const in, const vec_t* const v, int data_size, vec_t& w, vec_t& dw, int check_index).</p>
<p>in "network.h"</p>
<p>Regard for your reply</p>
<code>
std::cout << delta_by_bprop << "," << delta_by_numerical << std::endl;
</code>
<code>
return std::abs(delta_by_bprop - delta_by_numerical);
</code>
multi_layer:
0.00671147,0.0067115
0.00533005,0.00533007
0.0182457,0.0182457
-0.0235702,-0.0235702
0.00664365,0.00664369
0.0284786,0.0284786
0.00107073,0.00107072
-0.000101911,-0.000101918
0.0100989,0.0100989
-0.0134097,-0.0134097
-0.00467629,-0.00467633
-0.0180957,-0.0180957
-0.0183583,-0.0183583
0.00426405,0.00426408
-0.0112833,-0.0112833
-0.00614189,-0.00614191
0.00342017,0.00342018
-0.00606934,-0.00606935
0.0269247,0.0269246
0.0139144,0.0139144
-0.00357815,-0.00357814
0.0112781,0.0112781
-0.00777334,-0.00777336
0.00593223,0.00593225
0.00616244,0.00616243
-0.020952,-0.020952
0.00644011,0.00644009
0.0183032,0.0183032
0.00614859,0.00614859
0.0186381,0.0186381
-0.00680217,-0.00680216
0.00743266,0.00743268
0.0205075,0.0205075
-0.0223226,-0.0223226
0.00114486,0.00114488
-0.011131,-0.011131
0.00722955,0.00722955
-0.000360617,-0.0003606
-0.0194327,-0.0194327
0.00561516,0.00561513
-0.000800234,-0.000800249
0.0227468,0.0227468
0.00528352,0.00528353
-0.010697,-0.010697
-0.0151469,-0.0151469
-0.00570806,-0.00570806
0.0156926,0.0156926
-0.002475,-0.00247498
0.0113183,0.0113183
0.0284479,0.0284479
0.00671147,-0.0109401
gradient_check [ FAILED ]
<p>It always failed at i==50</p>
<p>in case GRAD_CHECK_ALL:
for (int i = 0; i < (int)w.size(); i++)</p>
<p>in "network.h"</p>

**Comment by dianyancao** [comment]
I have fix the bug in "fully_connected_layer.h".through modify the 0 to r.begin()
<code>vectorize::muladd(&curr_delta[**r.begin()**], prev_out[c], r.end() - r.begin(), &dW[c*out_size_ + r.begin()]);</code>

**Comment by nyanp** [comment]
@dianyancao 
Thank you very much!
It's a serious bug of tiny-cnn. I'll add your test case and fix the bug immediately.
